### PR TITLE
Fixed resource address in CosmosException. 

### LINKED ISSUE #16819: Expose API for showing current region served for the request - both read and write requests
* Expose API for showing current region served for the request - both read and write requests
* We already have this API in `ClientSideRequestStatistics`. We just need to expose on the `CosmosDiagnostics` and make it public. 
* `regionsContacted` API.

### PR DESCRIPTION
**Pull request by kushagraThapar**
* Fixed resourceAddress in CosmosException
* Added new API to expose regions contacted on CosmosDiagnostics
* TODO: @simplynaveen20 has a fix for regionsContacted API in ClientTelemetry PR, once that is in, it will fill up the regionsContacted for Gateway mode as well. Skipped that fix in this PR to avoid future conflicts

Sample resourceAddress and regionsContacted diagnostics: 

ResourceAddress sample DIRECT mode:

Exception is : NotFoundException{userAgent=azsdk-java-cosmos/4.9.0-beta.1 MacOSX/10.15.7 JRE/11.0.6, error={"Errors":["Resource Not Found. Learn more: https://aka.ms/cosmosdb-tsg-not-found"]}, resourceAddress='rntbd://cdb-ms-prod-westus2-fd31.documents.azure.com:14353/apps/e1145062-6507-492f-845f-f448225c36f6/services/55789df2-d46c-4435-9bc8-1d2ff3266202/partitions/8dedad13-a762-4b1d-b782-a38817f67371/replicas/132491704112230548s/, statusCode=404, message=["Resource Not Found. Learn more: https://aka.ms/cosmosdb-tsg-not-found"], ...

ResourceAddress sample GATEWAY mode:

resourceAddress='https://paged-flux-test-account-westus2.documents.azure.com:443/dbs/RxJava.SDKTest.SharedDatabase_20201110T150115_uRJ/colls/dbcd0b24-c2e4-41ed-b01e-2a0534ff3461/docs/cac7dadd-0f6f-4d7a-9490-c9494bf20019, statusCode=404, message=Entity with the specified id does not exist in the system. More info: https://aka.ms/cosmosdb-tsg-not-found, ...

RegionsContacted sample:
```
2020-11-06 13:32:32,013       [TestNG-method=directDiagnostics-1] INFO  TestSuiteBase - Regions contacted: [https://paged-flux-test-account-westus2.documents.azure.com:443/]
```

Closes #17279 
Closes #16819

### CODE REVIEW

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosException.java [hunk 1]
```diff
@@ -284,7 +284,7 @@ CosmosException setDiagnostics(CosmosDiagnostics cosmosDiagnostics) {
     @Override
     public String toString() {
         return getClass().getSimpleName() + "{" + "userAgent=" + USER_AGENT + ", error=" + cosmosError + ", resourceAddress='"
-                   + resourceAddress  +  "', requestUri='" + (requestUri != null ? requestUri.getURIAsString() : null) +  "', statusCode=" + statusCode + ", message=" + getMessage()
+                   + resourceAddress  +  ", statusCode=" + statusCode + ", message=" + getMessage()
                    + ", causeInfo=" + causeInfo() + ", responseHeaders=" + responseHeaders + ", requestHeaders="
```
Comment (xinlian12) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosException.java:288]: why we remove requestUri?
And I think currently we are mixing using resourceAddress and requestUri:
for example in RxGatewayStoreModel, we are passing resourceId to resourceAddress (XwJnANUYRi0=)
But in RntbdRequestManager, we are passing physical address to resourceAddress

Comment (kushagraThapar) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosException.java:288]: Removed requestUri from CosmosException string representation, as it is never set. 
In RxGatewayStoreModel, we are passing `request.getResourceAddress()`

Comment (kushagraThapar) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosException.java:288]: This is how it looks for gateway -> resourceAddress='dbs/RxJava.SDKTest.SharedDatabase_20201109T182237_GSh/colls/3d2be199-e23b-4af2-a755-510d7068ff97/docs/18b9d280-e899-4300-ae0c-5fe56411aa18, 

And for direct -> resourceAddress='rntbd://cdb-ms-prod-westus2-fd26.documents.azure.com:14376/apps/db2a149b-bdb4-4755-b26e-56d1a86cdd4e/services/8ce44d24-fe68-49ae-9784-94b322fa46a3/partitions/003a20cb-a187-4b3a-9bc9-306be0217c36/replicas/132494067072942319s/, 

which both are resource addresses.

Comment (xinlian12) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosException.java:288]: Thanks Kushagra.
I just wonder should we also capture the server info for gateway mode: (like following)
https://127.0.0.1:8081/dbs/RxJava.SDKTest.SharedDatabase_20201109T193921_erZ/colls/7d5337a3-30f7-46bb-b318-104a5dcfb721

Currently, the resourceAddress for gateway mode only contains the resourceId or full name, but not a complete uri.

Comment (kushagraThapar) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosException.java:288]: Good point, not sure how we can capture that, any ideas ?

Comment (xinlian12) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosException.java:288]: Currently we calculate the uri at RxGatewayStoreModel.performRequest, maybe we can add it in the requestContext?

**Review by FabianMeiswinkel — APPROVED** [review]
LGTM

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosException.java (hunk 1 above)
Comment (kushagraThapar) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosException.java:288]: Yes, I will try to do that, so we can re-use it, will update this thread once I have reached a conclusion :)

Comment (moderakh) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosException.java:288]: what about the direct mode? isn't that set in the direct mode? @kushagraThapar  could you please check.

**Review by moderakh — APPROVED** [review]
LGTM.

one question/comment please check if requestUri is used/set in the direct mode. if it is then we should keep it in the exception stacktrace. no?

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosException.java (hunk 1 above)
Comment (kushagraThapar) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosException.java:288]: @moderakh  - It is not used anywhere, other than at this code piece - https://github.com/Azure/azure-sdk-for-java/blob/master/sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/StoreReader.java#L808
Which I am not even sure, if that is correct or not, but don't plan to touch it  :)

Comment (kushagraThapar) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosException.java:288]: @xinlian12 @FabianMeiswinkel  - I have added the full resource address in gateway mode too, please check. 
Also updated the PR with sample.

Comment (xinlian12) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosException.java:288]: Thanks Kushagra, looks good to me~

**Review by xinlian12 — APPROVED** [review]
Thanks Kushagra~ LGTM

**Pull request merged**